### PR TITLE
Add published boolean field to region model

### PR DESCRIPTION
**Pull request by k9845**
Addresses
- #809 

## Changes
- Added published boolean field in Region model
- Added api for the publish `/api/v1/regions/{region.id}/publish/`

Mention related users here if any.
@samshara 
## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [ ] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: apps/geo/migrations/0038_region_published.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        ('geo', '0037_auto_20210503_0431'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='region',
+            name='published',
+            field=models.BooleanField(default=True),
+        ),
+    ]
```
Comment (thenav56) [inline apps/geo/migrations/0038_region_published.py]: Let's merge this two migration to one.

File: apps/geo/serializers.py [hunk 2]
```diff
@@ -64,6 +64,11 @@ def validate_project(self, project):
             raise serializers.ValidationError('Invalid project')
         return project.id
 
+    def validate(self, data):
+        if self.instance and self.instance.published == True:
```
Comment (thenav56) [inline apps/geo/serializers.py]:
```suggestion
        if self.instance and self.instance.is_published:
```

Also, add `is_published` to Meta.exclude. And add a new endpoint to publish the region.

File: apps/geo/serializers.py [hunk 3]
```diff
@@ -64,6 +64,11 @@ def validate_project(self, project):
...
+            raise serializers.ValidationError({'published': 'Published once set True cannot be revert back'})
```
Comment (thenav56) [inline apps/geo/serializers.py]:
```suggestion
            raise serializers.ValidationError('Published region can\'t be changed. Please contact Admin')
```
cc: @AdityaKhatri

File: apps/geo/models.py [hunk 4]
```diff
@@ -111,7 +114,7 @@ def can_modify(self, user):
                 project__regions=self,
                 member=user,
                 role__in=ProjectRole.get_admin_roles(),
-            ).exists())
+            ).exists())) and self.published == False
```
Comment (thenav56) [inline apps/geo/models.py]:
```suggestion
            ).exists())) and not self.is_published
```

File: apps/geo/models.py [hunk 5]
```diff
@@ -19,6 +21,7 @@ class Region(UserResource):
     code = models.CharField(max_length=10)
     title = models.CharField(max_length=255)
     public = models.BooleanField(default=True)
+    published = models.BooleanField(default=False)
```
Comment (thenav56) [inline apps/geo/models.py]: Additional work to do:
- Let's use `is_published`.
- `is_published` should be true for all current regions (we need to make sure that current regions are published)
- For new regions, `is_published` should be False by default and can be set by hitting an endpoint.

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Comment by thenav56** [comment]
@k9845 Add an endpoint to publish a region.
Also, tests are failing.

cc: @AdityaKhatri @suraj-ram-shrestha

File: apps/geo/serializers.py (hunk 3 above)
Comment (AdityaKhatri) [reply apps/geo/serializers.py]: Looks good to me

File: apps/geo/models.py [hunk 6]
```diff
@@ -111,9 +115,14 @@ def can_modify(self, user):
                 project__regions=self,
                 member=user,
                 role__in=ProjectRole.get_admin_roles(),
-            ).exists())
+            ).exists())) and not self.is_published
```
Comment (thenav56) [inline apps/geo/models.py]: Also, move not self.is_published check to the top.

File: apps/geo/models.py [hunk 7]
```diff
@@ -111,9 +115,14 @@ def can_modify(self, user):
...
         )
 
+    def can_publish(self, user):
+        if not Region.objects.filter(id=self, created_by=user):
+            raise ValidationError("Can be published by the user who created it")
+        return user
```
Comment (thenav56) [inline apps/geo/models.py]: Why is this here?

File: apps/geo/serializers.py [hunk 8]
```diff
@@ -50,7 +50,7 @@ class RegionSerializer(RemoveNullFieldsMixin,
 
     class Meta:
         model = Region
-        exclude = ('geo_options',)
+        exclude = ('geo_options', 'is_published')
```
Comment (thenav56) [inline apps/geo/serializers.py]: On second thought, remove is_published from here and add it as a read-only boolean field in the serializer.

File: apps/geo/views.py [hunk 9]
```diff
@@ -96,6 +99,23 @@ def get_intersects(self, request, pk=None, version=None):
             for id, geom in geoms
         ])
 
+    @action(
+        detail=True,
+        url_path='published',
```
Comment (thenav56) [inline apps/geo/views.py]: `publish` makes more sense.

```suggestion
        url_path='publish',
```

File: apps/geo/views.py [hunk 10]
```diff
@@ -96,6 +99,23 @@ def get_intersects(self, request, pk=None, version=None):
...
+        methods=['post'],
+        serializer_class = RegionSerializer,
```
Comment (thenav56) [inline apps/geo/views.py]: Flake8, please check.

```suggestion
        serializer_class=RegionSerializer,
```

File: apps/geo/views.py [hunk 11]
```diff
@@ -96,6 +99,23 @@ def get_intersects(self, request, pk=None, version=None):
...
+        permission_classes=[permissions.IsAuthenticated]
+    )
+    def get_published(self, request, pk=None, version=None):
+        region = self.get_object()
+        if not Region.objects.filter(id=region.id, created_by=request.user):
+            raise ValidationError('Can be published by user who created it')
+
+        region.is_published = True
+        region.save()
```
Comment (thenav56) [inline apps/geo/views.py]: Good habit to use updated_fields when we are changing single field only.

```suggestion
        region.save(update_fields=['is_published'])
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/geo/serializers.py [hunk 12]
```diff
@@ -47,7 +47,7 @@ class RegionSerializer(RemoveNullFieldsMixin,
         write_only=True,
         required=False,
     )
-
+    is_published = serializers.BooleanField(read_only=True)
```
Comment (thenav56) [inline apps/geo/serializers.py:50]: Flake8.  :expressionless: 

```suggestion
    is_published = serializers.BooleanField(read_only=True)

```

File: apps/geo/models.py [hunk 13]
```diff
@@ -111,9 +115,12 @@ def can_modify(self, user):
                 project__regions=self,
                 member=user,
                 role__in=ProjectRole.get_admin_roles(),
-            ).exists())
+            ).exists()))
         )
 
+    def can_publish(self, user):
+        return self.created_by == user
```
Comment (thenav56) [inline apps/geo/models.py]: Easy to read.

```suggestion
        return user == self.created_by
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Review by thenav56 — DISMISSED** [review]

**Comment by AdityaKhatri** [comment]
Lets merge this a bit later after we diverge as this is not something we'll be implementing in the old design.

**Review by annshress — DISMISSED** [review]

**Review by thenav56 — APPROVED** [review]

**Pull request merged**